Feature: connect Exante

alexander-myltsev opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

#### Description
<!--- Describe your changes in detail -->
Add Exante to LEAN

<!--- #### Related Issue -->
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->

#### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
Implement live trading using Exante brokerage account

<!---#### Requires Documentation Change -->
<!--- Please indicate if these changes will require updates to documentation, and if so, specify what changes are required -->

#### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Tests are at https://github.com/QuantConnect/Lean.Brokerages.Exante/tree/feature-exante_connect/QuantConnect.ExanteBrokerage.Tests .
The code has been tested both in demo and live mode.

#### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:
<!--- The following is a checklist of items that MUST be completed before a PR is accepted -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes. <!--- If not applicable, please explain why -->
- [x] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

<!--- Template inspired by https://www.talater.com/open-source-templates/#/page/99 -->

alexander-myltsev committed:
Move code from LEAN

alexander-myltsev committed:
Move LEAN projects to solution folder

alexander-myltsev committed:
Add `toolbox` project

alexander-myltsev committed:
Move template projects to solution folder

alexander-myltsev committed:
Remove `LimitIfTouchedOrder` from tests

since it isn't supported by Exante

alexander-myltsev committed:
Fix `ModifyOrder` expected response code

alexander-myltsev committed:
Fix Crypto symbol id

alexander-myltsev committed:
Implement brokerage GetExchange

alexander-myltsev committed:
Fix PlaceOrder

alexander-myltsev committed:
Minor fix to PlaceOrder

alexander-myltsev committed:
Fix to CancelOrder

alexander-myltsev committed:
Remove redundant code

alexander-myltsev committed:
Minor fixes

alexander-myltsev committed:
Remove redundant comments

alexander-myltsev committed:
Add opened orders to _orderMap

alexander-myltsev committed:
Make `GetOpenOrders` thread safe

alexander-myltsev committed:
Implement `GetAskPrice`

alexander-myltsev committed:
Switch to `ETH` symbol

alexander-myltsev committed:
Disable tests for not supported order types

alexander-myltsev committed:
Extract `ClientOptions` to helper class

alexander-myltsev committed:
Add `SymbolMapper` tests

alexander-myltsev committed:
Implement `GetHistory` and pass tests

alexander-myltsev committed:
Implement `CanPerformSelection`

alexander-myltsev committed:
Extract `GetCandlesHistory` to method

alexander-myltsev committed:
Add License

alexander-myltsev committed:
Move Exante toolbox project to root

alexander-myltsev committed:
Minor fixes to symbol mapper

alexander-myltsev committed:
Implement Toolbox

alexander-myltsev committed:
Delete template projects

Martin-Molinero committed:
A few tweaks. Addressing reviews

Martin-Molinero reviewed (changes requested):
Cool! Thanks for the contribution 🙌✨  
Leaving a few comments and pushing a commit with a couple of small tweaks

alexander-myltsev committed:
Those files should in LEAN

https://github.com/QuantConnect/Lean.Brokerages.Exante/pull/2#discussion_r736688080

alexander-myltsev committed:
Explicitly ignore in test the missing functionality

https://github.com/QuantConnect/Lean.Brokerages.Exante/pull/2#discussion_r736751961

alexander-myltsev committed:
Use Market and SecurityType args when download data with toolbox

https://github.com/QuantConnect/Lean.Brokerages.Exante/pull/2#discussion_r736755783

alexander-myltsev committed:
Minor fix

https://github.com/QuantConnect/Lean.Brokerages.Exante/pull/2#discussion_r736760636

alexander-myltsev committed:
Remove redundant null test

https://github.com/QuantConnect/Lean.Brokerages.Exante/pull/2#discussion_r736766751

alexander-myltsev committed:
Minor fix: remove order from internal order map if it's filled

alexander-myltsev committed:
Use `CurrencyPairUtil.DecomposeCurrencyPair` while converting to Exante symbol

https://github.com/QuantConnect/Lean.Brokerages.Exante/pull/2#discussion_r736767171

alexander-myltsev committed:
Expand var name

https://github.com/QuantConnect/Lean.Brokerages.Exante/pull/2#discussion_r736771890

alexander-myltsev committed:
Remove the IDataQueueUniverseProvider

https://github.com/QuantConnect/Lean.Brokerages.Exante/pull/2#discussion_r736773413

alexander-myltsev committed:
Move `IsCanonical` to `CanSubscribe`

https://github.com/QuantConnect/Lean.Brokerages.Exante/pull/2#discussion_r736773921

alexander-myltsev committed:
Minor fixes

alexander-myltsev committed:
Properly handle `Subscribe` to symbol errors

https://github.com/QuantConnect/Lean.Brokerages.Exante/pull/2#discussion_r736775672

alexander-myltsev committed:
Define feed stream callback methods

https://github.com/QuantConnect/Lean.Brokerages.Exante/pull/2#discussion_r736776296

alexander-myltsev committed:
Remove duplicated `_subscribedTickersStreamSubscriptions`

https://github.com/QuantConnect/Lean.Brokerages.Exante/pull/2#discussion_r736777224

alexander-myltsev committed:
Minor fix

https://github.com/QuantConnect/Lean.Brokerages.Exante/pull/2#discussion_r736778702

alexander-myltsev commented on QuantConnect.ExanteBrokerage/ExanteBrokerage.cs:
```diff
@@ -0,0 +1,845 @@
+            }
+
+            var updateResult = true;
+            foreach (var bi in order.BrokerId.Skip(1))
```
The only scenario (confirmed by Exante support) when REST API creates multiple orders in response is when placing a limit order with take-profit and stop-loss (one order for the limit order, one connected stop order if any, and one connected limit order with a price if any). It looks as follows in the terminal:

![image](https://user-images.githubusercontent.com/574267/139499205-044ce89e-36da-4b97-9ac2-902c2b1ca0d6.png)

The API call to reproduce it is as follows:

```c#
exanteClient.PlaceOrderAsync(accountId, "ETH.USD", ExanteOrderType.Limit, ExanteOrderSide.Buy,
            quantity: 1m, ExanteOrderDuration.GoodTillCancel, limitPrice: 4160m, 
            takeProfit: 4300m, stopLoss: 4100).Result;
```

And since LEAN doesn't support such kind of orders (AFAIK), I handle only the scenario when order placement produces only a single order.

alexander-myltsev replied:
Added comments at becac1dd0c848779251af45d96ac6e881d542919. Feel free to close if you're convinced.

alexander-myltsev commented on QuantConnect.ExanteBrokerage/ExanteBrokerage.cs:
```diff
@@ -0,0 +1,845 @@
+                    case OrderType.StopLimit:
+                        var stopLimitOrder = (StopLimitOrder)order;
+                        exanteOrder = Client.ModifyOrder(
+                            Guid.Parse(order.BrokerId.First()),
```
I had same reason as described at https://github.com/QuantConnect/Lean.Brokerages.Exante/pull/2#discussion_r739521819

alexander-myltsev replied:
Added comments at becac1dd0c848779251af45d96ac6e881d542919. Feel free to close if you're convinced.

alexander-myltsev committed:
Add comments explaining why `Skip(1)`/`First`

https://github.com/QuantConnect/Lean.Brokerages.Exante/pull/2#discussion_r736772000
https://github.com/QuantConnect/Lean.Brokerages.Exante/pull/2#discussion_r736772653

alexander-myltsev commented on QuantConnect.ExanteBrokerage/ExanteBrokerage.cs:
```diff
@@ -0,0 +1,845 @@
+                return;
+            }
+
+            Thread.Sleep(1_000); // Need to wait for `Client.GetTransactions(...)`
```
Please, let me know if you know the better way to achieve the same result.

Martin-Molinero replied:
Ah right, generally we shouldn't Thread.Sleep anywhere. Delaying an order event an entire second seems excessive I think, could we rely on the fee model instead? any deviation should be solved by the cash sync performed daily by Lean instead

alexander-myltsev replied:
WDYT of an alternative of launching `while (true)` loop with `100ms` delay between calls until it gets the result?

alexander-myltsev replied:
@Martin-Molinero Martin, ⬆️  ?

jaredbroad replied:
@alexander-myltsev  We approximate some live trading for speed, fetching fees is one of those places as it can be easily modeled most of the time. @Martin-Molinero is saying to use the ExanteBrokerageFeeModel to avoid the call entirely. 

We are slowly systematically removing `thread.sleep()` from the LEAN engine and replacing it with ManualResetEvent's where required - -but this isn't a location which needs one. We just need to use the fee model.

alexander-myltsev replied:
Fixed by 5b51d2a28c494285c0601a0e39959116956dcf5f. @Martin-Molinero Martin, please check

alexander-myltsev commented:
> I see your using a dependency `Exante.Net` https://www.nuget.org/packages/Exante.Net/ which seems very new and not official, wondering how much benefit does it bring... my initial though is suggesting against adding it as a dependency. How much effort would it take to avoid it?

What are the alternatives? There are no official .NET clients from Exante I'm aware of. 

Exante provides only the Swagger stuff. I used the Swagger-generated C# code at the beginning of my journey. But Jared said it's better to use Exante.NET that he found himself. Also, the Swagger-generated code is usually of lower quality and requires manual tweaks to fit the needs. Also also, I wonder if it supports handling incoming feed (orders updates, ticks) facilities.

I'm afraid I don't have the time capacity to start and switch to an Exante C# client that repeats Exante.NET. Please, consider cloning it and publishing artifacts on your side.

Martin-Molinero commented on QuantConnect.ExanteBrokerage/ExanteBrokerage.cs:
```diff
@@ -0,0 +1,845 @@
+            var bids = exanteTickShort.Bid.ToList();
+            var asks = exanteTickShort.Ask.ToList();
+            return new Tick(time, symbol, "", "",
+                bids.IsNullOrEmpty() ? decimal.Zero : bids[0].Size,
+                bids.IsNullOrEmpty() ? decimal.Zero : bids[0].Price,
+                asks.IsNullOrEmpty() ? decimal.Zero : asks[0].Size,
+                asks.IsNullOrEmpty() ? decimal.Zero : asks[0].Price);
```
My concern is `bids` and `asks` are a list and you are only using the first entry and dismissing the rest of the information, so I wonder why? Is there value in the rest of the data in the list which we should be using, my guess is there is

alexander-myltsev replied:
I'm still figuring out it from Exante.

alexander-myltsev replied:
The response from the Exante support:

> In relation to your question, please, let us inform you about the following:
upon requesting an instrument quote stream, you will be receiving a single result for a bid and a single result for an ask.

I have only seen only single element in the array. I suppose this might be for some future functionality.

Martin-Molinero commented on QuantConnect.ExanteBrokerage/ExanteBrokerageFactory.cs:
```diff
@@ -0,0 +1,131 @@
+                throw new Exception(string.Join(System.Environment.NewLine, errors));
+            }
+
+            var clientOptions = CreateExanteClientOptions(clientId, applicationId, sharedKey, platformTypeStr);
```
I guess it would be more generic if it were ExanteSettings or something similar and we just passed that instance around, will make the API cleaner I think?

alexander-myltsev replied:
fixed by 60e49eae6b34581c57e29eaf6feb96800344ea22

Martin-Molinero commented on QuantConnect.ExanteBrokerage/ExanteBrokerage.cs:
```diff
@@ -0,0 +1,845 @@
+        /// <summary>
+        /// Returns the brokerage account's base currency
+        /// </summary>
+        public override string AccountBaseCurrency => Currencies.USD;
```
> I suppose it's a client-side preference.

Cool 👍  I guess it's okay as is then

Martin-Molinero commented on QuantConnect.ExanteBrokerage.Tests/ExanteBrokerageSymbolMapperTests.cs:
```diff
@@ -0,0 +1,52 @@
+        {
+            var mapper = SymbolMapper();
+
+            var symbol = mapper.GetLeanSymbol("EUR/USD", SecurityType.Forex, Market.Oanda);
```
> Which Market should be for that ticker?

Which market is EUR/USD in Exante? I guess it's probably their own right? Meaning it would be Exante Market I believe. It's not like exante is connected to Oanda and trading to it?

alexander-myltsev replied:
There are two sections available at the Exante terminal: Cash and Forex. The difference is explained by Exante support as follows:


> Please be informed that the "Forex" tab is intended for trading currency pairs, that is, it offers trading instruments. At the same time, the "Cash" tab (Currency) is an internal list of pairs intended for currency conversion within your account.
>
>"Currency" section are used for money conversion which is stored on your account. For example, having dollars, you can convert part of them, or whole amount in Euros, Rubles, and so on at the current market rate.
>
> In relation to your question regarding Crypto we would like to let you know that crypto is not a centralized instrument, there is no single market price provider.
We are cooperating with different exchanges who provide different rates.
> Operations on Cash (.exante) instruments do not appear in Open Positions, no PnL is calculated on them. These instruments are purely for cash conversion.
>
>Clearing on different markets is performed by different prime brokers and Tier One banks. At the same time we are not a clearing member and work through our partners. We have several counterparties for each type of instrument. Orders and deals can be viewed in the reports of exchanges and our liquidity providers in case of OTC trade. As we operate under an STP model, we don't trade against our clients and do not have a dealing desk that modifies clients' trades. We do have manual trading services for bonds and exotic financial instruments execution.
>
> Unfortunately more detailed information we disclose only confidentially, on conditions of no further disclosure. It's an international practice.

Do you still think introducing `Market.Exante` is the correct way?

alexander-myltsev replied:
@Martin-Molinero ping

alexander-myltsev replied:
@Martin-Molinero Martin?

Martin-Molinero replied:
Sorry for the delay @alexander-myltsev

> We are cooperating with different exchanges who provide different rates.

If they provide information about the exchange each tick is coming from, we could determine the actual exchange and translate it to a Lean market. Not sure what's coming in for currency pairs?

![image](https://user-images.githubusercontent.com/18473240/158388998-e74e6c00-ad0c-4bca-bfaa-2213485bd730.png)

If currencies don't have exchange information, seems to me we will need to use `Market.Exante`. The cons of adding `Market.Exante` though is how does it translate to backtesting, each market in lean has it's own data set. We would need to add the historical data for the currency pairs (which is what is done for Oanda). If required this could potentially be handled by the data downloader.

alexander-myltsev replied:
There are mappings between Exante and LEAN markets at https://github.com/QuantConnect/Lean.Brokerages.Exante/blob/54fbbebce61ed3724de5abb6fa4e10e0370f32c6/QuantConnect.ExanteBrokerage/ExanteSymbolMapper.cs#L59

As I said previously, Exante provides its local "Cash" and "Forex" for currencies. If it's "Cash", then "exchange" is "EXANTE" (e.g. `USD/SGD.EXANTE`). If it's "Forex", then exchange is "E.FX" (e.g. `USD/SGD.E.FX`). How should it implemented then?

alexander-myltsev replied:
@Martin-Molinero ping ⬆️

Martin-Molinero commented on QuantConnect.ExanteBrokerage/ExanteBrokerage.Utility.cs:
```diff
@@ -0,0 +1,410 @@
+                        exchange == Market.CBOE ||
+                        exchange == Market.SGX)
+                    {
+                        market = exchange;
```
You don't seem to have access to the underlying here really but something like `symbol.HasUnderlying ? symbol.Underlying.ID.Market` maybe, could maybe use the security type too to be smart about it 🤔

alexander-myltsev replied:
fixed by 73b7ae4f7c28358958169bf9eded8b6ae3b68db0

Martin-Molinero commented on QuantConnect.ExanteBrokerage/ExanteSymbolMapper.cs:
```diff
@@ -0,0 +1,244 @@
+        public ExanteSymbolLocal GetExanteSymbol(Symbol symbol)
+        {
+            if (string.IsNullOrWhiteSpace(symbol?.Value))
+                throw new ArgumentException($"Invalid symbol: {(symbol == null ? "null" : symbol.ToString())}");
```
> Is the symbol itself null check is required? Or does it always come non-null?

Suggest checking the callers context, don't think it should be null, can also follow other brokerages pattern 👍

alexander-myltsev replied:
fixed by 7e1ea9d319ef0ecd935bda177e31f734cc445f6b

Martin-Molinero commented on QuantConnect.ExanteBrokerage/ExanteSymbolMapper.cs:
```diff
@@ -0,0 +1,244 @@
+                    {
+                        if (market != tickerToExchange[sym])
+                        {
+                            Log.Error($"Symbol {sym} occurs on two exchanges: {tickerToExchange[sym]} {market}");
```
> Also, how do you propose to distinguish CRMD traded on NASDAQ and NYSE?

For Lean, US equities are all under the same Market.USA, the security is the same for Lean. Although you can route to one or another exchange if supported by the brokerage see `OrderProperties.cs` in Lean

> I doubt I have all the information at that point to construct Lean Symbol to get its ID.

Maybe can add the market/securityType as part of the key for example. You could still potentially have clashes for options and futures?

alexander-myltsev replied:
fixed by 541b9907249f11398f5db7d087d66c95ed299787

alexander-myltsev committed:
Introduce `ExanteBrokerageOptions`

alexander-myltsev committed:
Fix: if the underlying is an US equity the market of the option for Lean is also USA

https://github.com/QuantConnect/Lean.Brokerages.Exante/pull/2#discussion_r736760092

alexander-myltsev committed:
Add missing XML docs

alexander-myltsev committed:
Symbol mapper: don't log error on duplicates when both are from US exchange

Fixes https://github.com/QuantConnect/Lean.Brokerages.Exante/pull/2#discussion_r736768918

alexander-myltsev committed:
Use `ExanteBrokerageFeeModel` to avoid `Thread.Sleep`

Fixes https://github.com/QuantConnect/Lean.Brokerages.Exante/pull/2#discussion_r816235736

Martin-Molinero commented on QuantConnect.ExanteBrokerage/ExanteBrokerage.cs:
```diff
@@ -0,0 +1,839 @@
+
+            var security = _securityProvider.GetSecurity(symbol);
+            var orderFeeParameters =  new OrderFeeParameters(security, order);
+            var fee = _feeModel.GetOrderFee(orderFeeParameters);
```
Cool! Believe should directly use the `security.FeeModel`, instead of a general `_feeModel` 👍

alexander-myltsev replied:
I looked at your `security.FeeModel` related changes and found the problems as follows. 

1. I found that I initialised `_securityProvider` with `QuantConnect.Tests.Brokerages.SecurityProvider` from tests. The class from the production is `SecurityPortfolioManager`. But it seems like overkill to create it. What's the proper way to get security from symbol?

2. Speaking of `security.FeeModel`. I check all the usages of [FeeModel setter](https://github.com/QuantConnect/Lean/blob/83f9499b4a8faf3a1344b6f22284a35ba91d2e1b/Common/Securities/Security.cs#L186-L186). And I found it fills explicitly, but not by security constructor. Why then you suggest to use `security.FeeModel` instead of class field instance `_feeModel: FeeModel`?

Martin-Molinero replied:
> What's the proper way to get security from symbol?

There are plenty of example in Lean code to checkout, for example https://github.com/QuantConnect/Lean/blob/master/Tests/Brokerages/Kraken/KrakenFeeModelTests.cs#L38

> Why then you suggest to use security.FeeModel instead of class field instance _feeModel: FeeModel?

Because users can overwrite it and use what ever feel model they want, depending on the asset for example, shouldn't be a global fee model

alexander-myltsev replied:
I see that `new Security` constructor at the line you pointed creates `Security` with `InteractiveBrokersFeeModel`. Also, `KrakenFeeModelTests` is a test. I'm still not following how to change `FeeModel`. I was searching for a general security provider like at 3deba219ffd5d62daba949d022726d028f575781 which provides `ExanteFeeModel`. Could you explain how to do it properly?

Martin-Molinero replied:
Can follow example at https://github.com/QuantConnect/Lean/blob/30d7fb042bbd1da407c829efc4dbe936b282793a/Brokerages/Oanda/OandaBrokerage.cs#L67 & https://github.com/QuantConnect/Lean/blob/30d7fb042bbd1da407c829efc4dbe936b282793a/Brokerages/Oanda/OandaBrokerageFactory.cs#L98 the factory can pass in the security provider and the brokerage just needs to ask for the security, with the security the correct fee model to use can be recovered.
In the SetJob use case, like https://github.com/QuantConnect/Lean/blob/30d7fb042bbd1da407c829efc4dbe936b282793a/Brokerages/Oanda/OandaBrokerage.cs#L263 , null is passed in because the brokerage is used as a IDataQueueHandler, not like a brokerage for trading, so it should never be used in those code paths

alexander-myltsev replied:
OK. I attempted to fix as you told at f868937345c5d226925281813349981df6a11af8

Martin-Molinero commented on QuantConnect.ExanteBrokerage/ExanteBrokerage.cs:
```diff
@@ -0,0 +1,839 @@
+
+            foreach (var kline in history)
+            {
+                yield return new TradeBar()
```
For quote tick type should be creating a `QuoteBar` here instead?

alexander-myltsev replied:
Fixed by 32a47686d709e8eef44f9838b6030184499d7407

Martin-Molinero replied:
> DataType = MarketDataType.TradeBar,

Should use the QuoteBar and TradeBar constructors instead of the public setters, to avoid these issues.
Does exante provide bid/ask bar information for the quoteBar?

alexander-myltsev replied:
Attempted to fix it at 372cb7dda86f2f05c70ce792139ddc2b1faa1871. I didn't manage to create `QuoteBar` from constructor because I'm not sure where to get bid/ask. I expect you asked about bid/ask of QuoteBar for that reason. `ExanteCandle` doesn't have bid/ask

```c#
public class ExanteCandle
{
  /// <summary>Candle close price</summary>
  public Decimal Close { get; set; }
  /// <summary>Candle open price</summary>
  public Decimal Open { get; set; }
  /// <summary>Candle low price</summary>
  public Decimal Low { get; set; }
  /// <summary>Candle high price</summary>
  public Decimal High { get; set; }
  /// <summary>
  /// Total volume for specified period. Appears and required only for trade candle request
  /// </summary>
  public Decimal? Volume { get; set; }
  /// <summary>Candle date</summary>
  [JsonProperty(PropertyName = "timestamp")]
  [JsonConverter(typeof (TimestampConverter))]
  public DateTime Date { get; set; }
}
```

There are methods [getQuoteLast](https://api-live.exante.eu/api-docs/#operation/getQuoteLast) and [getQuoteStream](https://api-live.exante.eu/api-docs/#operation/getQuoteStream) that return bid/ask. But that means extra REST-API call that is not what you'd like.

Martin-Molinero replied:
> ExanteCandle

This seems like the Lean `TradeBar` not `QuoteBar`. 

> There are methods [getQuoteLast](https://api-live.exante.eu/api-docs/#operation/getQuoteLast) and [getQuoteStream](https://api-live.exante.eu/api-docs/#operation/getQuoteStream) that return bid/ask. But that means extra REST-API call that is not what you'd like.

The problem isn't the api call if required, but those endpoints don't return historical data? 
If exante doesn't return quote information, that's okay it should `yield return` detecting that case, see similar handling:
https://github.com/QuantConnect/Lean/blob/master/Brokerages/GDAX/GDAXBrokerage.cs#L339
 https://github.com/QuantConnect/Lean/blob/master/Brokerages/Bitfinex/BitfinexBrokerage.cs#L338

alexander-myltsev replied:
Well, I don't know the method of Exante API that returns quote information. Seems like there isn't such method. Let's leave `TradeBar` only then.

Fixed at 11e79095f7355b43cf5ef10a4169ecd8d02b6b8b

Martin-Molinero commented on QuantConnect.ExanteBrokerage/ExanteBrokerage.cs:
```diff
@@ -0,0 +1,839 @@
+        /// <param name="job">Job we're subscribing for</param>
+        public void SetJob(LiveNodePacket job)
+        {
+        }
```
Will need a couple of tweaks/updates to support the new usage of multiple data queue handlers, see examples at https://github.com/QuantConnect/Lean/pull/6047

alexander-myltsev replied:
@Martin-Molinero Martin, check those changes 7b3b56fafdf5c6063936e08d9923a4f3c7bdf206, please.

Martin-Molinero replied:
> new ExanteBrokerageFactory().CreateBrokerage(job, null);

Will create a new instance and current instance will not really be initialized? Can see examples at https://github.com/QuantConnect/Lean/pull/6047

alexander-myltsev replied:
Why not initialised? Take a look at https://github.com/QuantConnect/Lean.Brokerages.Exante/blob/7b3b56fafdf5c6063936e08d9923a4f3c7bdf206/QuantConnect.ExanteBrokerage/ExanteBrokerageFactory.cs#L68-L95

It extracts all the necessary parameters from the `LiveNodePacket job`. I choose not to use `Initialise` method because it requires some fields to drop `readonly` modifier.

Another thing I noticed is that there is extra routine on connecting it. Since `ExanteClient` is REST HTTP, I decided drop the check. But pro forma if you ask I added it at 439f1e7ada21a023b198b6e418724a0ad90b31f2.

Martin-Molinero replied:
> new ExanteBrokerageFactory().CreateBrokerage(job, null);

This code is from a method in an instance of an already created brokerage, but doing `new ExanteBrokerageFactory().CreateBrokerage(job, null)` you are creating a new factory and an entirely new brokerage instance here. Should follow examples at https://github.com/QuantConnect/Lean/pull/6047

alexander-myltsev replied:
I've got your point. Fixed by f868937345c5d226925281813349981df6a11af8

Martin-Molinero replied:
The current implementation doesn't seem to be following the pattern yet, for example `ExanteBrokerage()` vs `BitfinexBrokerage()` please review

Martin-Molinero committed:
Address selfreview

Martin-Molinero committed:
Merge branch 'master' of github.com:QuantConnect/Lean.Brokerages.Exante into HEAD

Martin-Molinero commented on QuantConnect.ExanteBrokerage/ExanteSymbolMapper.cs:
```diff
@@ -0,0 +1,303 @@
+
+        public ExanteSymbolMapper(
+            ExanteClientWrapper client,
+            HashSet<string> supportedCryptoCurrencies)
```
I think we should not hard code any ticker, they will get outdated really fast. Could we replace this by an API call? Or just let it fail when trying to add a currency that isn't supported at runtime?
Symbols API
![image](https://user-images.githubusercontent.com/18473240/158389103-19616a87-99c4-4685-9395-e8dc501c4315.png)

alexander-myltsev replied:
I tried the approach you suggest. There are problems as follows. There are cryptos:

![image](https://user-images.githubusercontent.com/574267/159991674-39563e33-4c6b-4dba-a4fd-c74f66f4bc68.png)

But those don't have explicit exchange in ticker:

![image](https://user-images.githubusercontent.com/574267/159991793-55e604ce-f526-48da-8a6d-5148713861a9.png)

Though, I can't filter symbols by exchange. Some time ago cryptos were under `ExanteSymbolType.Fund` as stated at https://github.com/d-ugarov/Exante.Net/issues/1. But Exante moved it from that type. And I don't know where are they now. 

In total, I don't know using which properties to filter all the symbols from the API.

alexander-myltsev replied:
I asked Exante support once again. They just don't answer how to get the list via API, but simply send me to REST API documentation.

Experimentally I found out the code to fetch the exact same list:

```c#
List<ExanteSymbol> cryptoCurrencies =
                    _exanteClient
                        .GetSymbolsByTypeAsync(ExanteSymbolType.CFD)
                        .SynchronouslyAwaitTaskResult()
                        .Data
                        .Where(x => x.Exchange == "USD")
                        .ToList();
```

Previously they were under `Fund` symbol type. I don't know why it under `CFD` and with `USD` `Exchange`. Either don't know how long it would be there. Do you think I should replace hard-coded list with that code?

Martin-Molinero replied:
CFD sounds like [Contract for Differences](https://www.investopedia.com/terms/c/contractfordifferences.asp) which for Lean is it's own asset type with it's own contract specifications (can't trade CFDs in the US. Can see examples at https://github.com/QuantConnect/Lean/blob/master/Data/symbol-properties/symbol-properties-database.csv . Is that the case for Exante? Seems like the market would be Exante for those assets like in those examples and we don't really need any list in the brokerage implementation?

alexander-myltsev replied:
@Martin-Molinero Martin, the Exante support would not understand the question about eligibility of trading CFDs in USA. If you somehow related to US, please check if there are crypto currencies available in terminal in demo mode.

Martin-Molinero replied:
Suggest not supporting crypto trading on exante for now and ship the rest first, which is a lot already, we can go back and review it in depth 👍 edit: please to cleanup the crypto unrequired changes/features/code for now

alexander-myltsev replied:
Fixed by 3580b3af440564c934575b3b988bebcbd122886e

Martin-Molinero committed:
Minor tweak

Martin-Molinero reviewed (commented):
Hey @alexander-myltsev!
Sorry for the delay here, thank you for your work 👍 

Posted a few comments, and pushed a couple of minor commits

alexander-myltsev committed:
Return QuoteBar for quote tick type

Fixes https://github.com/QuantConnect/Lean.Brokerages.Exante/pull/2#discussion_r826465519

alexander-myltsev committed:
Minor fixes

alexander-myltsev committed:
Implement `SetJob`

Fixes https://github.com/QuantConnect/Lean.Brokerages.Exante/pull/2#discussion_r826466586

Martin-Molinero commented on QuantConnect.ExanteBrokerage/ExanteBrokerage.cs:
```diff
@@ -0,0 +1,850 @@
+                            Time = kline.Date,
+                            Symbol = request.Symbol,
+                            Value = kline.Close,
+                            DataType = MarketDataType.TradeBar,
```
> DataType = MarketDataType.TradeBar,

Should use the QuoteBar and TradeBar constructors instead of the public setters, to avoid these issues.
Does exante provide bid/ask bar information for the quoteBar?

alexander-myltsev committed:
Implement `ExanteSecurityProvider`

alexander-myltsev committed:
Fix Quote/TradeBar type

https://github.com/QuantConnect/Lean.Brokerages.Exante/pull/2#discussion_r842183075

alexander-myltsev committed:
Connect Exante brokerage if necessary in `SetJob`

alexander-myltsev committed:
Update `Exante.Net`

Martin-Molinero commented on QuantConnect.ExanteBrokerage/ExanteSecurityProvider.cs:
```diff
@@ -0,0 +1,81 @@
+       
+        private static Security CreateSecurity(Symbol symbol)
+        {
+            var security = new Security(
```
The brokerage shouldn't be creating securities, this should just be a method in a test class maybe? not production

alexander-myltsev replied:
Fixed by b318c4b046bdbee4aa10fdfa2ed425b113b84c21

alexander-myltsev committed:
Move `ExanteSecurityProvider` to test suite

alexander-myltsev committed:
Properly implement `SetJob`

Fixes https://github.com/QuantConnect/Lean.Brokerages.Exante/pull/2#discussion_r854570561

alexander-myltsev committed:
Return only `TradeBar` on `GetCandlesHistory` since Exante API doesn't support quotes

Fixes https://github.com/QuantConnect/Lean.Brokerages.Exante/pull/2#discussion_r854573874

alexander-myltsev committed:
Remove crypto currencies support from Exante brokerage

https://github.com/QuantConnect/Lean.Brokerages.Exante/pull/2#discussion_r854580871

alexander-myltsev committed:
Replace crypto currencies with `EURUSD` in tests

https://github.com/QuantConnect/Lean.Brokerages.Exante/pull/2#discussion_r854580871

Martin-Molinero commented:
Some extra todos:
- CI should be green
- Requires net6 update
- Sanity test live deployment
    - Long running: deploy dummy algo trading now and then for a week
    - Connection outage: deploy dummy algo subscription to > 50 symbols and cut internet connection, wait 1 min and reconnect, subscriptions should be reestablished/data flowing. Should also place a trade after re connection to make sure order events are also flowing

alexander-myltsev committed:
Update dependencies

alexander-myltsev committed:
Ignore most of tests

alexander-myltsev committed:
Add `QC.Tests` project reference

alexander-myltsev committed:
Update github checkout

alexander-myltsev committed:
Update `Newtonsoft.Json`

alexander-myltsev committed:
Add ENV vars to `Run Tests`

alexander-myltsev committed:
Implement basic Exante client test

alexander-myltsev committed:
Pass Exante tests in Debug mode

alexander-myltsev committed:
Minor fixes

alexander-myltsev committed:
Add HKEX market. Use `GEELY` symbol for tests

alexander-myltsev committed:
Add `RateGate` limiter

alexander-myltsev committed:
Return all tests

alexander-myltsev committed:
github workflow: add `dotnet restore`. And switch to `Release`

alexander-myltsev committed:
Remove unused project dependency
